extensions/parquet/table/src/main/java/io/deephaven/parquet/table/layout/ParquetFlatPartitionedLayout.java

```diff
@@ -5,42 +5,61 @@
 
 import io.deephaven.engine.table.impl.locations.TableDataException;
 import io.deephaven.engine.table.impl.locations.impl.TableLocationKeyFinder;
+import io.deephaven.parquet.base.ParquetFileReader;
+import io.deephaven.parquet.base.ParquetFileReader.ParquetFileReaderException;
+import io.deephaven.parquet.base.util.LocalFSChannelProvider;
 import io.deephaven.parquet.table.location.ParquetTableLocationKey;
 import org.jetbrains.annotations.NotNull;
 
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.DirectoryStream;
-import java.nio.file.Files;
+import java.nio.file.FileSystems;
 import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.Consumer;
 
-import static io.deephaven.parquet.table.ParquetTableWriter.PARQUET_FILE_EXTENSION;
-
 /**
  * Parquet {@link TableLocationKeyFinder location finder} that will discover multiple files in a single directory.
  */
 public final class ParquetFlatPartitionedLayout implements TableLocationKeyFinder<ParquetTableLocationKey> {
 
+    private static ParquetTableLocationKey locationKey(Path path) {
+        return new ParquetTableLocationKey(path.toFile(), 0, null);
+    }
+
     private final File tableRootDirectory;
+    private final Map<Path, ParquetTableLocationKey> cache;
 
     /**
      * @param tableRootDirectory The directory to search for .parquet files.
      */
     public ParquetFlatPartitionedLayout(@NotNull final File tableRootDirectory) {
         this.tableRootDirectory = tableRootDirectory;
+        cache = new HashMap<>();
     }
 
     public String toString() {
         return ParquetFlatPartitionedLayout.class.getSimpleName() + '[' + tableRootDirectory + ']';
     }
 
     @Override
-    public void findKeys(@NotNull final Consumer<ParquetTableLocationKey> locationKeyObserver) {
-        try (final DirectoryStream<Path> parquetFileStream =
-                Files.newDirectoryStream(tableRootDirectory.toPath(), "*" + PARQUET_FILE_EXTENSION)) {
+    public synchronized void findKeys(@NotNull final Consumer<ParquetTableLocationKey> locationKeyObserver) {
+        try (final DirectoryStream<Path> parquetFileStream = FileSystems.getDefault().provider()
+                .newDirectoryStream(tableRootDirectory.toPath(), ParquetFileHelper::fileNameMatches)) {
             for (final Path parquetFilePath : parquetFileStream) {
-                locationKeyObserver.accept(new ParquetTableLocationKey(parquetFilePath.toFile(), 0, null));
+                ParquetTableLocationKey locationKey = cache.get(parquetFilePath);
+                if (locationKey == null) {
+                    try {
+                        new ParquetFileReader(parquetFilePath.toString(), new LocalFSChannelProvider());
```

Comment:
I know I asked for this, but you're correct that it's expensive. We should probably discuss a little bit.

Reply:
Agreed.

Reply:
I think we can check for the trailer more cheaply than by constructing a reader. If we're going to construct the reader, it would be better to get full benefits of doing so (caching-wise), and use the correct channel provider; you can do this by optimistically constructing the PTLK and calling `io.deephaven.parquet.table.location.ParquetTableLocationKey#getFileReader`.

Reply:
Yeah - I was looking into that a bit. I wasn't sure if there's a lot of extraneous overhead due to the CachedChannelProvider / TrackedSeekableChannelsProvider / TrackedFileHandleFactory in that path.

Additionally, this code path explicitly catches IOException and elevates to TableDataException. We can extract that inner part I guess, so we can catch the IOException ourselves.